diff --git a/lib/pure/distros.nim b/lib/pure/distros.nim
index 8d4fbc857..26f092712 100644
--- a/lib/pure/distros.nim
+++ b/lib/pure/distros.nim
@@ -191,18 +191,6 @@ template detectOs*(d: untyped): bool =
   ## enum value.
   detectOsImpl(Distribution.d)
 
-when not defined(nimble):
-  var foreignDeps: seq[string] = @[]
-
-proc foreignCmd*(cmd: string; requiresSudo = false) =
-  ## Registers a foreign command to the intern list of commands
-  ## that can be queried later.
-  let c = (if requiresSudo: "sudo " else: "") & cmd
-  when defined(nimble):
-    nimscriptapi.foreignDeps.add(c)
-  else:
-    foreignDeps.add(c)
-
 proc foreignDepInstallCmd*(foreignPackageName: string): (string, bool) =
   ## Returns the distro's native command line to install 'foreignPackageName'
   ## and whether it requires root/admin rights.
@@ -243,19 +231,3 @@ proc foreignDepInstallCmd*(foreignPackageName: string): (string, bool) =
     result = ("pkgman install " & p, true)
   else:
     result = ("brew install " & p, false)
-
-proc foreignDep*(foreignPackageName: string) =
-  ## Registers 'foreignPackageName' to the internal list of foreign deps.
-  ## It is your job to ensure the package name
-  let (installCmd, sudo) = foreignDepInstallCmd(foreignPackageName)
-  foreignCmd installCmd, sudo
-
-proc echoForeignDeps*() =
-  ## Writes the list of registered foreign deps to stdout.
-  for d in foreignDeps:
-    echo d
-
-when false:
-  foreignDep("libblas-dev")
-  foreignDep "libfoo"
-  echoForeignDeps()
