diff --git a/compiler/cgen.nim b/compiler/cgen.nim
index 625e48328..6cf47cd08 100644
--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -1359,8 +1359,8 @@ proc genMainProc(m: BModule) =
       "}$N$N"
 
     GenodeNimMain =
-      "extern Genode::Env *nim_runtime_env;$N" &
-      "extern void nim_component_construct(Genode::Env*);$N$N" &
+      "extern \"C\" Genode::Env *nim_runtime_env;$N" &
+      "extern \"C\" void nim_component_construct(Genode::Env*);$N$N" &
       NimMainBody
 
     ComponentConstruct =
